### ISSUE TITLE
Demo application will not link demo application under windows, with nng.lib built as static library

### ISSUE THREAD

**Issue by rheslar123** [nanomsg/nng#1639]
NNG / Nanomsg compilation under windows

Created by [Robert Heslar](https://medacuity.atlassian.net/wiki/people/63408f2dfedc6169aeda678a?ref=confluence&src=profilecard), 
with a template
Just a moment ago3 min read
Analytics
this is instructions i have used to build the NNG library from sources

visual studio 2017 and above is required for these steps

:blue_book: Instructions

clone the repo: [GitHub - nanomsg/nng: nanomsg-next-generation -- light-weight brokerless messaging](https://github.com/nanomsg/nng.git) 

open a visual studio command prompt, and navigate to the build directory

run cmake .., from the command line, this will create a visual studio solution named nng.sln

open the nng.sln, in visual studio

in visual studio, rebuild the solution, we now have a nng.lib file, for use with a NNG / Nanomsg Application

Highlight important information in a panel like this one. To edit this panel's color or style, select one of the options in the menu.

:clipboard: Related articles
this shows the instructions used to build the NNG asyc application demo, found in the NNG demos folder

create the folder path as below, so cmake can find the library and header files for the NNG asyc application demo:

copy the nng includes into the include path above, the include directory in the NNG asyc application demo path:

copy the nng.lib file from the nng/build/Debug directory(from steps in the above article), to the NNG asyc application lib directory 

for this NNG application, have this CMakelist.text file:

CMakeLists.txt
26 Jan 2023, 10:09 AM
6) open a visual studio command prompt, and navigate to the NNG asyc application demo build directory

7) navigate to the  NNG asyc application demo build directory path

8 run cmake .., form the NNG asyc application demo build directory

9) a visual studio solution will be created with the name: nng-asyncdemo.sln

10) open the nng-asyncdemo.sln in visual studio

11) run rebuild all, in the nng-asyncdemo solution, in visual studio

 

problems: (added nng.lib from steps before (compiling nng source under windows)

when compiling the NNG asyc application demo visual studio we now get the following linker errors:

NNG application missing link symbols:

2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_close referenced in function server_function
2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_listen referenced in function server_function
2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_listener_create referenced in function run_server
2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_listener_start referenced in function run_server
2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_listener_get_addr referenced in function run_server
2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_strerror referenced in function fatal
2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_send referenced in function server_function
2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_recv referenced in function recive_function1
2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_free referenced in function recive_function1
2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_surveyor0_open referenced in function server_function
[CMakeLists.txt](https://github.com/nanomsg/nng/files/10511376/CMakeLists.txt)
[CMakeLists.txt](https://github.com/nanomsg/nng/files/10511378/CMakeLists.txt)

**Comment by gdamore** [comment]
Yeah, the problem here is an age old one of how Windows DLLs work.  There are probably ways we could make this a little less painful. 

The NNG_STATIC_LIB macro should be set when building a static library, which causes the symbols to be renamed  -- without that we default to using the dynamic version of symbol naming, and you get these link failures.

This is only a problem on Windows, and I'm not sure how to do a better job of making this easy for folks using DLLs or using static libraries.  Right now the default favors DLL users.

**Comment by gdamore** [comment]
If some Windows / CMake GURU comes along, maybe they can offer some wisdom to make this better.

**Comment by alawn-wang** [comment]
> NNG / Nanomsg compilation under windows
> 
> Created by [Robert Heslar](https://medacuity.atlassian.net/wiki/people/63408f2dfedc6169aeda678a?ref=confluence&src=profilecard), with a template Just a moment ago3 min read Analytics this is instructions i have used to build the NNG library from sources
> 
> visual studio 2017 and above is required for these steps
> 
> 📘 Instructions
> 
> clone the repo: [GitHub - nanomsg/nng: nanomsg-next-generation -- light-weight brokerless messaging](https://github.com/nanomsg/nng.git)
> 
> open a visual studio command prompt, and navigate to the build directory
> 
> run cmake .., from the command line, this will create a visual studio solution named nng.sln
> 
> open the nng.sln, in visual studio
> 
> in visual studio, rebuild the solution, we now have a nng.lib file, for use with a NNG / Nanomsg Application
> 
> Highlight important information in a panel like this one. To edit this panel's color or style, select one of the options in the menu.
> 
> 📋 Related articles this shows the instructions used to build the NNG asyc application demo, found in the NNG demos folder
> 
> create the folder path as below, so cmake can find the library and header files for the NNG asyc application demo:
> 
> copy the nng includes into the include path above, the include directory in the NNG asyc application demo path:
> 
> copy the nng.lib file from the nng/build/Debug directory(from steps in the above article), to the NNG asyc application lib directory
> 
> for this NNG application, have this CMakelist.text file:
> 
> CMakeLists.txt 26 Jan 2023, 10:09 AM 6) open a visual studio command prompt, and navigate to the NNG asyc application demo build directory
> 
> 7. navigate to the  NNG asyc application demo build directory path
> 
> 8 run cmake .., form the NNG asyc application demo build directory
> 
> 9. a visual studio solution will be created with the name: nng-asyncdemo.sln
> 10. open the nng-asyncdemo.sln in visual studio
> 11. run rebuild all, in the nng-asyncdemo solution, in visual studio
> 
> problems: (added nng.lib from steps before (compiling nng source under windows)
> 
> when compiling the NNG asyc application demo visual studio we now get the following linker errors:
> 
> NNG application missing link symbols:
> 
> 2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_close referenced in function server_function 2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_listen referenced in function server_function 2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_listener_create referenced in function run_server 2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_listener_start referenced in function run_server 2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_listener_get_addr referenced in function run_server 2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_strerror referenced in function fatal 2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_send referenced in function server_function 2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_recv referenced in function recive_function1 2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_free referenced in function recive_function1 2>survey.obj : error LNK2019: unresolved external symbol __imp_nng_surveyor0_open referenced in function server_function [CMakeLists.txt](https://github.com/nanomsg/nng/files/10511376/CMakeLists.txt) [CMakeLists.txt](https://github.com/nanomsg/nng/files/10511378/CMakeLists.txt)

if you want to use static lib in windows, maybe you should add a pre-compile instruction for your pre-processor such like this 'add_compile_definitions(NNG_STATIC_LIB)'

**Comment by gdamore** [comment]
I *suspect* that this was possibly a consequence of using downrev CMake versions.  I've bumped them all up to 3.15 for the demos.  Please let me know if this is still a problem.

THe workaround is indeed to add NNG_STATIC_LIB if it is missing, but that should be done automatically when you import the project.

**Comment by gdamore** [comment]
So I understand this a lot better now.

First, yes, we needed to use newer CMake.
Second, we needed to expose find_packages in our config, and use the CONFIG variant of find_package.  (this was a bug, pr in flight right now to fix -- only impacted MbedTLS and WolfSSL because I already did that for Threads.
Third, you have to *install* NNG somewhere so find_package of *it* can work.  I think that's reasonable -- because that's what the demos are supposed to be showing.

**Comment by gdamore** [comment]
The fix for the second issue was just merged, and I've tested that I can indeed build and run the various demos under Windows, even with static libraries.  Closing.
